File: src/modules/comments/comment.service.ts
import { db } from "@/db";
import { comments, users } from "@/db/schema";
import { eq, and, desc, sql } from "drizzle-orm";

export interface CommentWithUser {
  id: string;
  decisionId: string;
  userId: string;
  content: string;
  isInternal: boolean | null;
  createdAt: Date | null;
  updatedAt: Date | null;
  userName: string;
  userRole: string;
}

export const commentService = {
  async create(input: {
    decisionId: string;
    userId: string;
    content: string;
    isInternal: boolean;
  }) {
    const [comment] = await db
      .insert(comments)
      .values({
        decisionId: input.decisionId,
        userId: input.userId,
        content: input.content,
        isInternal: input.isInternal,
      })
      .returning();

    return comment;
  },

  async getByDecision(
    decisionId: string,
    includeInternal: boolean
  ): Promise<CommentWithUser[]> {
    const conditions = [eq(comments.decisionId, decisionId)];

    if (!includeInternal) {
      conditions.push(eq(comments.isInternal, false));
    }

    const rows = await db
      .select({
        id: comments.id,
        decisionId: comments.decisionId,
        userId: comments.userId,
        content: comments.content,
        isInternal: comments.isInternal,
        createdAt: comments.createdAt,
        updatedAt: comments.updatedAt,
        userName: users.name,
        userRole: users.role,
      })
      .from(comments)
      .innerJoin(users, eq(comments.userId, users.id))
      .where(and(...conditions))
      .orderBy(comments.createdAt);

    return rows;
  },

  async delete(id: string): Promise<void> {
    await db.delete(comments).where(eq(comments.id, id));
  },

  async count(decisionId: string): Promise<number> {
    const [result] = await db
      .select({ count: sql<number>`count(*)::int` })
      .from(comments)
      .where(eq(comments.decisionId, decisionId));

    return result?.count ?? 0;
  },

  async getById(id: string) {
    const [comment] = await db
      .select()
      .from(comments)
      .where(eq(comments.id, id))
      .limit(1);

    return comment ?? null;
  },
};

File: backend/src/DiscretionaryPowers.Infrastructure/Services/CommentService.cs
using DiscretionaryPowers.Application.DTOs.Comments;
using DiscretionaryPowers.Domain.Entities;
using DiscretionaryPowers.Domain.Interfaces;
using DiscretionaryPowers.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace DiscretionaryPowers.Infrastructure.Services;

public class CommentService(AppDbContext db) : ICommentService
{
    public async Task<Comment> Create(Guid decisionId, Guid userId, string content, bool isInternal)
    {
        var comment = new Comment
        {
            Id = Guid.NewGuid(),
            DecisionId = decisionId,
            UserId = userId,
            Content = content,
            IsInternal = isInternal,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };

        db.Comments.Add(comment);
        await db.SaveChangesAsync();

        return comment;
    }

    public async Task<IReadOnlyList<Comment>> GetByDecision(Guid decisionId, bool includeInternal = false)
    {
        var q = db.Comments
            .AsNoTracking()
            .Include(c => c.User)
            .Where(c => c.DecisionId == decisionId);

        if (!includeInternal)
            q = q.Where(c => !c.IsInternal);

        return await q
            .OrderByDescending(c => c.CreatedAt)
            .ToListAsync();
    }

    public async Task Delete(Guid id)
    {
        var comment = await db.Comments.FindAsync(id);
        if (comment is not null)
        {
            db.Comments.Remove(comment);
            await db.SaveChangesAsync();
        }
    }

    public async Task<int> Count(Guid decisionId)
    {
        return await db.Comments.CountAsync(c => c.DecisionId == decisionId);
    }

    public async Task<Comment?> GetById(Guid id)
    {
        return await db.Comments
            .AsNoTracking()
            .Include(c => c.User)
            .FirstOrDefaultAsync(c => c.Id == id);
    }

    public static CommentResponse MapToResponse(Comment c) => new()
    {
        Id = c.Id,
        DecisionId = c.DecisionId,
        UserId = c.UserId,
        Content = c.Content,
        IsInternal = c.IsInternal,
        UserName = c.User?.Name ?? "Unknown",
        UserRole = c.User?.Role.ToString().ToLowerInvariant() ?? "unknown",
        CreatedAt = c.CreatedAt,
        UpdatedAt = c.UpdatedAt,
    };
}
